x-pack/plugins/security/server/authentication/providers/saml.ts

```diff
@@ -457,19 +474,22 @@ export class SAMLAuthenticationProvider extends BaseAuthenticationProvider {
     try {
       // This operation should be performed on behalf of the user with a privilege that normal
       // user usually doesn't have `cluster:admin/xpack/security/saml/prepare`.
-      const { id: requestId, redirect } = await this.options.client.callAsInternalUser(
-        'shield.samlPrepare',
-        { body: { realm: this.realm } }
-      );
+      const {
+        id: requestId,
+        redirect: redirectWithoutRelayState,
+      } = await this.options.client.callAsInternalUser('shield.samlPrepare', {
+        body: { realm: this.realm },
+      });
+
+      // HACK: we manually add a `RelayState` query string parameter with URL to redirect user to after
+      // successful SAML handshake since Elasticsearch doesn't support it yet. We may break something
+      // here eventually if we keep this workaround for too long.
+      const redirect = `${redirectWithoutRelayState}&RelayState=${encodeURIComponent(nextURL)}`;
```

Comment:
__note:__ there was the idea to somehow encrypt `RelayState` so that it can't be altered, but I doubt it's a big a problem if we don't. Also keeping `RelayState` as a plain path#fragment would allow us to specify any URL for IdP initiated login, e.g. like described in `Auth0` [docs](https://auth0.com/docs/protocols/saml/saml-configuration/special-configuration-scenarios/idp-initiated-sso).